akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala

```diff
@@ -93,17 +92,28 @@ private[io] final class AsyncDnsResolver(
           log.debug("{} cached {}", mode, resolved)
           sender() ! resolved
         case None =>
-          resolveWithResolvers(name, mode, resolvers)
-            .map { resolved =>
-              if (resolved.records.nonEmpty) {
-                val minTtl = (positiveCachePolicy +: resolved.records.map(_.ttl)).min
-                cache.put((name, mode), resolved, minTtl)
-              } else if (negativeCachePolicy != Never) cache.put((name, mode), resolved, negativeCachePolicy)
-              log.debug(s"{} resolved {}", mode, resolved)
-              resolved
+          val replyTo = sender()
+          resolveWithResolvers(name, mode, resolvers).onComplete {
+            _ match {
+              case Success(resolved) =>
+                if (resolved.records.nonEmpty) {
+                  val minTtl = (positiveCachePolicy +: resolved.records.map(_.ttl)).min
+                  cache.put((name, mode), resolved, minTtl) // thread-safe structure, safe in callback
+                } else if (negativeCachePolicy != Never) {
+                  cache.put((name, mode), resolved, negativeCachePolicy) // thread-safe structure, safe in callback
```

Comment:
Maybe updating the cache inside the actor is better?

Reply:
If it was going to just be a `Map`, definitely.  In this case, since we're going through the expense of using a thread-safe structure, may as well exercise that :)

This does have the benefit of quickly allowing a successful resolution to immediately update the cache for requests which might be in the mailbox now (without having to do a custom mailbox).  Since the only `cache.get` is right when we receive a `DnsProtocol.Resolve`, it's not much of a win (it might be a slight win under heavy load to check the cache when making follow-up queries).